File: tools/test/create-broker-release-fixture.py
#!/usr/bin/env python3
"""Create small oliphaunt-broker release-shaped assets for SDK checks."""

from __future__ import annotations

import argparse
from pathlib import Path

from release_fixture_utils import write_checksum_manifest, write_tar_gz, write_zip


def broker_entries(target: str, executable: str) -> dict[str, bytes]:
    return {
        executable: b"#!/bin/sh\necho oliphaunt-broker release fixture\n",
        "manifest.properties": (
            b"schema=oliphaunt-broker-release-assets-v1\n"
            b"product=oliphaunt-broker\n"
            + f"target={target}\n".encode()
            + f"binary={executable}\n".encode()
        ),
    }


def write_fixture_assets(asset_dir: Path, version: str) -> None:
    asset_dir.mkdir(parents=True, exist_ok=True)
    executable_modes = {"bin/oliphaunt-broker": 0o755, "bin/oliphaunt-broker.exe": 0o755}

    for target in ["macos-arm64", "linux-x64-gnu", "linux-arm64-gnu"]:
        write_tar_gz(
            asset_dir / f"oliphaunt-broker-{version}-{target}.tar.gz",
            broker_entries(target, "bin/oliphaunt-broker"),
            executable_modes,
        )

    write_zip(
        asset_dir / f"oliphaunt-broker-{version}-windows-x64-msvc.zip",
        broker_entries("windows-x64-msvc", "bin/oliphaunt-broker.exe"),
        executable_modes,
    )
    write_checksum_manifest(asset_dir, f"oliphaunt-broker-{version}-release-assets.sha256")


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--asset-dir", required=True, help="directory to write release-shaped assets into")
    parser.add_argument("--version", required=True, help="oliphaunt-broker version to encode in asset names")
    return parser.parse_args()


def main() -> int:
    args = parse_args()
    write_fixture_assets(Path(args.asset_dir).resolve(), args.version)
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: tools/test/create-broker-release-fixture.mjs
#!/usr/bin/env bun
import { createHash } from 'node:crypto';
import fs from 'node:fs/promises';
import os from 'node:os';
import path from 'node:path';

import { stageBrokerDependencyLicenses } from '../release/broker-dependency-license-contract.mjs';
import { stageReleaseNotices } from '../release/release-notices.mjs';

import {
  elfFixture,
  machoFixture,
  parseCommonArgs,
  windowsPeFixture,
  writeChecksumManifest,
  writeEntriesArchive,
} from './release-fixture-utils.mjs';

function brokerBinary(target) {
  if (target === 'macos-arm64') {
    return machoFixture({ platform: 1, minos: [11, 0, 0] });
  }
  if (target === 'linux-x64-gnu') {
    return elfFixture({ machine: 62, requiredVersions: ['GLIBC_2.17'] });
  }
  if (target === 'linux-arm64-gnu') {
    return elfFixture({ machine: 183, requiredVersions: ['GLIBC_2.17'] });
  }
  throw new Error(`unsupported broker release fixture target ${target}`);
}

async function carrierLegalEntries(target) {
  const stage = await fs.mkdtemp(path.join(os.tmpdir(), 'oliphaunt-broker-fixture-legal-'));
  await fs.chmod(stage, 0o755);
  try {
    stageReleaseNotices(stage, { profile: 'broker' });
    stageBrokerDependencyLicenses(stage, target);
    const entries = {};
    async function walk(directory, relative = '') {
      for (const entry of await fs.readdir(directory, { withFileTypes: true })) {
        const member = relative ? `${relative}/${entry.name}` : entry.name;
        const file = path.join(directory, entry.name);
        if (entry.isDirectory()) {
          await walk(file, member);
        } else if (entry.isFile()) {
          entries[member] = await fs.readFile(file);
        } else {
          throw new Error(`unexpected broker legal fixture entry ${file}`);
        }
      }
    }
    await walk(stage);
    return entries;
  } finally {
    await fs.rm(stage, { recursive: true, force: true });
  }
}

async function brokerEntries(target, executable) {
  return {
    ...await carrierLegalEntries(target),
    [executable]: brokerBinary(target),
    'manifest.properties': [
      'schema=oliphaunt-broker-release-assets-v1',
      'product=oliphaunt-broker',
      `target=${target}`,
      `binary=${executable}`,
      '',
    ].join('\n'),
  };
}

async function windowsBrokerEntries() {
  const runtimeName = 'vcruntime140.dll';
  const executable = windowsPeFixture({ imports: ['VCRUNTIME140.dll'] });
  const runtime = windowsPeFixture({ imports: ['KERNEL32.dll'] });
  const digest = createHash('sha256').update(runtime).digest('hex');
  return {
    ...await carrierLegalEntries('windows-x64-msvc'),
    'bin/oliphaunt-broker.exe': executable,
    [`bin/${runtimeName}`]: runtime,
    'bin/windows-vc-runtime.sha256': `${digest}  ${runtimeName}\n`,
    'manifest.properties': [
      'schema=oliphaunt-broker-release-assets-v1',
      'product=oliphaunt-broker',
      'target=windows-x64-msvc',
      'binary=bin/oliphaunt-broker.exe',
      `windowsVcRuntimeDlls=${runtimeName}`,
      '',
    ].join('\n'),
  };
}

async function writeFixtureAssets(assetDir, version) {
  await fs.mkdir(assetDir, { recursive: true });
  const executableModes = {
    'bin/oliphaunt-broker': 0o755,
    'bin/oliphaunt-broker.exe': 0o755,
  };

  for (const target of ['macos-arm64', 'linux-x64-gnu', 'linux-arm64-gnu']) {
    await writeEntriesArchive(
      path.join(assetDir, `oliphaunt-broker-${version}-${target}.tar.gz`),
      await brokerEntries(target, 'bin/oliphaunt-broker'),
      executableModes,
    );
  }

  await writeEntriesArchive(
    path.join(assetDir, `oliphaunt-broker-${version}-windows-x64-msvc.zip`),
    await windowsBrokerEntries(),
    executableModes,
  );
  await writeChecksumManifest(assetDir, `oliphaunt-broker-${version}-release-assets.sha256`);
}

const { assetDir, version } = parseCommonArgs(
  Bun.argv.slice(2),
  'Create small oliphaunt-broker release-shaped assets for SDK checks.',
);
await writeFixtureAssets(assetDir, version);
